s2fft/precompute_transforms/construct.py

```diff
@@ -193,39 +369,103 @@ def wigner_kernel(
         forward (bool, optional): Whether to provide forward or inverse shift.
             Defaults to False.
 
+        mode (str, optional): Whether to use FFT approach or manually compute each element.
+            {"auto", "manual", "irfft"}. Defaults to "auto" which will detect the
+            most appropriate recursion given the parameter configuration.
+
         using_torch (bool, optional): Desired frontend functionality. Defaults to False.
 
     Returns:
         np.ndarray: Transform kernel for Wigner transform.
     """
+    if mode.lower() == "fft" and sampling.lower() not in ["mw", "mwss", "dh"]:
+        mode = "manual"
+        warn(
+            f"Fourier based recursion is not valid for {sampling} sampling. "
+            + "Defaulting to auto mode."
+        )
+    # Determine operational mode automatically.
+    # - Can only use the FFT approach when uniformly sampling in theta.
+    # - FFT approach is only more efficient when N <= L/Log(L) roughly.
+    if mode.lower() == "auto":
+        if sampling.lower() in ["mw", "mwss", "dh"]:
+            mode = "fft" if N <= int(L / np.log(L)) else "manual"
+        else:
+            mode = "manual"
+
     n_start_ind = N - 1 if reality else 0
     n_dim = N if reality else 2 * N - 1
 
+    # Determine theta locations for forward vs inverse transform.
     if forward and sampling.lower() in ["mw", "mwss"]:
         sampling = "mwss"
         thetas = samples.thetas(2 * L, "mwss")
     else:
         thetas = samples.thetas(L, sampling, nside)
 
+    # Range values which need only be defined once.
+    m_value = np.arange(-L + 1, L)
+    n = np.arange(n_start_ind - N + 1, N)
     dl = np.zeros((n_dim, len(thetas), L, 2 * L - 1), dtype=np.float64)
-    for n in range(n_start_ind - N + 1, N):
-        for t, theta in enumerate(thetas):
-            for el in range(abs(n), L):
-                ind = n if reality else N - 1 + n
-                dl[ind, t, el] = recursions.turok.compute_slice(theta, el, L, n, False)
 
+    # GL and HP sampling ARE NOT uniform in theta therefore CANNOT be calculated
+    # using the Fourier decomposition of Wigner d-functions. Instead they must be
+    # manually bootstrapped from the recursion.
```

Comment:
Let's not say bootsrapped.  That has a precise statistical meaning.  I would just say "directly computed"...

Reply:
changed to calculated.